docs/concepts/tx-lifecycle.md

```diff
@@ -0,0 +1,196 @@
+
+#### DeliverTx
+
+The `DeliverTx` ABCI function defined in [`baseapp`](./baseapp.md) does the bulk of the state change work: it is run for each transaction in the block in sequential order as committed to during consensus. Under the hood, `DeliverTx` is almost identical to `CheckTx` but calls the [`runTx`](./baseapp.md#runtx) function in deliver mode instead of check mode. Instead of using their `checkTxState` or `queryState`, full-nodes select a new copy, `deliverTxState`, to deliver transactions:
+
+* **Decoding:** `Tx` is decoded.
+* **Initializations:** The `Context`, `Multistore`, and `GasMeter` are initialized.
+* **Checks:** The full-nodes call `validateBasicMsgs` and the `AnteHandler` again. This second check happens because they may not have seen the same transactions during the Addition to Mempool stage and a malicious proposer may have included invalid transactions.
+* **Route and Handler:** The extra step is to run `runMsgs` to fully execute each `Msg` within the transaction. Since `tx` may have messages from different modules, `baseapp` needs to know which module to find the appropriate Handler. Thus, the [`Route`](./msg-tx.md#route) function is called to retrieve the route name and find the `MsgHandler`.
+* **MsgHandler:** The `MsgHandler`, a step up from `AnteHandler`, is responsible for executing each message's actions and causes state changes to persist in `deliverTxState`. It is defined within a `Msg`'s module and writes to the appropriate stores within the module.
+* **Gas:** While the transaction is being delivered, a `GasMeter` is used to keep track of how much gas is left for each transaction; GasUsed is deducted from it and returned in the `Response`. Deferred functions run at the end of `runTx` for the purpose of checking how much gas was used and outputting `GasUsed` and `GasWanted` in the `Response`. If `GasMeter` has run out or something else goes wrong, it appropriately errors or panics.
+
+If there are any failed state changes resulting from `tx` being invalid or `GasMeter` running out, `tx` processing terminates and any state changes are reverted. Any leftover gas is returned to the user.
+
+#### EndBlock
+
+[`EndBlock`](./app-anatomy.md#beginblocker-and-endblocker) is always run at the end and is useful for automatic function calls or changing governance/validator parameters. No transactions are handled here.
+
+#### Commit
+
+The application's `Commit` method is run in order to finalize the state changes made by executing this block's transactions. A new state root should be sent back to serve as a merkle proof for the state change. Any application can inherit Baseapp's [`Commit`](./baseapp.md#commit) method; it syncs all the states by writing the `deliverTxState` into the application's internal state. Moving forward, `checkTxState`, `deliverTxState`, and `queryState` will all start afresh from the most recently committed state in order to be consistent and reflect the changes.
```

Comment:
"this block" which block?

Reply:
sent back where?

Reply:
"Moving forward, `checkTxState`, `deliverTxState`, and `queryState` will all start afresh from the most recently committed state in order to be consistent and reflect the changes." when?